File: solutions/java/secret-handshake/2/src/main/java/HandshakeCalculator.java
import java.util.*;

class HandshakeCalculator {
    private final Map<Integer, Signal> mapping = new HashMap<>();

    public HandshakeCalculator() {
        mapping.put(1, Signal.WINK);
        mapping.put(2, Signal.DOUBLE_BLINK);
        mapping.put(4, Signal.CLOSE_YOUR_EYES);
        mapping.put(8, Signal.JUMP);
    }

    public List<Signal> calculateHandshake(int number) {
        List<Signal> signals = new ArrayList<>();

        mapping.forEach((key, value) -> {
            if ((number & key) == key) {
                signals.add(value);
            }
        });

        if ((number & 16) == 16) {
            Collections.reverse(signals);
        }

        return signals;
    }
}

File: solutions/kotlin/secret-handshake/2/src/main/kotlin/HandshakeCalculator.kt
object HandshakeCalculator {
    private fun Int.has(signalIndex: Int) = (this shr signalIndex) and 0x1 == 1

    fun calculateHandshake(i: Int): List<Signal> {
        val signals = Signal.values().mapIndexedNotNull { index, signal ->
            if (i.has(index)) signal else null
        }

        return if (i.has(Signal.values().size)) signals.reversed() else signals
    }
}
